### ISSUE TITLE
ManyToMany::synchronize provides invalid syntax for PostgreSQL

### ISSUE THREAD

**Issue by mathysie** [mako-framework/framework#229]
For my application I have a PostgreSQL-database with two tables, `profiles` and `studies`, with a ManyToMany-relation between them.

After editing a profile, I update the relation between profiles and studies using:
```$profile->studies()->synchronize($_POST['studies'] ?? []);```

Here `$_POST['studies']` is an array with id's of studies needed to be linked to the profile.

This provides the following stack trace:
```
[2017-10-12 17:24:16] mako.ERROR: PDOException: SQLSTATE[42601]: Syntax error: 7 ERROR:  syntax error at or near ")" 
LINE 1: ...files_studies" WHERE "profile_id" = $1 AND "studie_id" IN ()
                                                                      ^ in /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/connections/Connection.php:516
Stack trace:
#0 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/connections/Connection.php(516): PDOStatement->execute()
#1 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/connections/Connection.php(549): mako\database\connections\Connection->execute(Array)
#2 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/query/Query.php(1495): mako\database\connections\Connection->queryAndCount('DELETE FROM "pr...', Array)
#3 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/midgard/relations/ManyToMany.php(275): mako\database\query\Query->delete()
#4 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/midgard/relations/ManyToMany.php(310): mako\database\midgard\relations\ManyToMany->unlink(Array)
#5 /var/www/deleidscheflesch.nl/app/controllers/Bedrijven.php(473): mako\database\midgard\relations\ManyToMany->synchronize(Array)
#6 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/syringe/Container.php(476): app\controllers\Bedrijven->profielOpslaan('37')
#7 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/http/routing/Dispatcher.php(183): mako\syringe\Container->call(Array, Array)
#8 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/http/routing/Dispatcher.php(211): mako\http\routing\Dispatcher->executeController(Object(app\controllers\Bedrijven))
#9 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/http/routing/Dispatcher.php(227): mako\http\routing\Dispatcher->executeAction()
#10 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/onion/Onion.php(113): mako\http\routing\Dispatcher->mako\http\routing\{closure}(Object(mako\http\Request), Object(mako\http\Response))
#11 /var/www/deleidscheflesch.nl/app/routing/middleware/AuthorizationRequiredMiddleware.php(70): mako\onion\Onion->mako\onion\{closure}(Object(mako\http\Request), Object(mako\http\Response))
#12 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/onion/Onion.php(128): app\routing\middleware\AuthorizationRequiredMiddleware->execute(Object(mako\http\Request), Object(mako\http\Response), Object(Closure))
#13 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/onion/Onion.php(163): mako\onion\Onion->mako\onion\{closure}(Object(mako\http\Request), Object(mako\http\Response))
#14 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/http/routing/Dispatcher.php(228): mako\onion\Onion->peel(Object(Closure), Array)
#15 /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/application/web/Application.php(55): mako\http\routing\Dispatcher->dispatch()
#16 /var/www/deleidscheflesch.nl/public/index.php(17): mako\application\web\Application->run()
#17 {main} {"exception":"[object] (PDOException(code: 42601): SQLSTATE[42601]: Syntax error: 7 ERROR:  syntax error at or near \")\" 
LINE 1: ...files_studies\" WHERE \"profile_id\" = $1 AND \"studie_id\" IN ()
                                                                      ^ at /var/www/deleidscheflesch.nl/vendor/mako/framework/src/mako/database/connections/Connection.php:516)"} 
```
Is there a way to provide valid syntax for PostgreSQL?

**Comment by mathysie** [comment]
This exception is thrown when nothing regarding studiesz is changed or whess studies are added. The removal of studies does not throw an exception.

**Comment by freost** [comment]
It seems that the issue is caused by an empty array. Wouldn't something like this fix the issue?

```php
if(!empty($_POST['studies'])
{
    $profile->studies()->synchronize($_POST['studies']);
}
```

**Comment by mathysie** [comment]
Unfortunately, the error also arises with non-empty arrays. If, for example, a profile is linked to studies 1, 2 and 3 and `$_POST['studies'] = ['1', '2', '3']` holds, the exception is thrown. If a profile is linked to studies 1 and 2 with the same `$_POST`, the exception is thrown too. Only when I remove a study from my profile, there is no exception.

**Comment by freost** [comment]
Ok. I'll have to do some testing and probably write some more unit/integration tests.

**Comment by mathysie** [comment]
I was just about to ask whether `$query->in($this->getJunctionKey(), $keys);` in the `unlink()` method is incorrectly being accessed due to `$id` being not null.

**Comment by freost** [comment]
Fixed in 5.2.8

**Comment by mathysie** [comment]
My issue is resolved, thanks!
